pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -125,14 +237,255 @@ func localityLbEndpointsFromInstances(instances []*model.ServiceInstance) []endp
+				// TODO: once the deps are updated, log the ErrorCode if set (missing in current version)
+				if edsDebug {
+					log.Infof("EDS: ACK %s %s %s", node, discReq.VersionInfo, con.Clusters)
+				}
+				continue
+			} else {
+				if len(con.Clusters) > 0 {
+					// Should not happen
+					log.Infof("EDS REQ repeated %s %v/%v %v raw: %s ",
+						node, clusters2, con.Clusters, peerAddr, discReq.String())
+				}
+				// Initial request
+				if edsDebug {
+					log.Infof("EDS REQ %s %v %v raw: %s ",
+						node, clusters2, peerAddr, discReq.String())
+				}
+			}
+			if len(clusters2) > 0 {
+				con.Clusters = clusters2
+			}
+
+			for _, c := range con.Clusters {
+				s.addEdsCon(c, node, con)
+			}
+
+		case <-con.pushChannel:
+			if len(con.Clusters) == 0 {
+				continue
+			}
+		}
+
+		if len(con.Clusters) > 0 {
+			response := s.endpoints(s.mesh, con.Clusters)
+			err := stream.Send(response)
+			if err != nil {
+				return err
+			}
+
+			if edsDebug {
+				log.Infof("EDS RES for %s %q clusters %v, Response: \n%s\n",
+					node, peerAddr, con.Clusters, response.String())
+			}
+		} else {
+			if edsDebug {
+				log.Infof("EDS empty clusters %v \n", peerAddr)
+			}
+		}
+	}
+}
+
+// EdsPushAll implements old style invalidation, generated when any rule or endpoint changes.
+// Primary code path is from v1 discoveryService.clearCache(), which is added as a handler
+// to the model ConfigStorageCache and Controller.
+func EdsPushAll() {
+	if edsDebug {
+		log.Infoa("EDS cache reset")
+	}
+	edsClusterMutex.Lock()
+	// Create a temp map to avoid locking the add/remove
+	tmpMap := map[string]*EdsCluster{}
+	for k, v := range edsClusters {
+		tmpMap[k] = v
+	}
+	version++
+	edsClusterMutex.Unlock()
+
+	for clusterName, edsCluster := range tmpMap {
+		updateCluster(clusterName, edsCluster)
+		for _, edsCon := range edsCluster.EdsClients {
+			edsCon.pushChannel <- true
+		}
+	}
+}
+
+// Edsz implements a status and debug interface for EDS.
+// It is mapped to /debug/edsz on the monitor port (9093).
+func Edsz(w http.ResponseWriter, req *http.Request) {
+	if req.Form.Get("debug") != "" {
+		edsDebug = req.Form.Get("debug") == "1"
+		return
+	}
+	if req.Form.Get("push") != "" {
+		EdsPushAll()
+	}
+	data, err := json.Marshal(edsClusters)
+	if err != nil {
+		_, _ = w.Write([]byte(err.Error()))
+		return
+	}
+
+	_, _ = w.Write(data)
+}
+
+// addEdsCon will track the eds connection, for push and debug
+func (s *DiscoveryServer) addEdsCon(clusterName string, node string, connection *EdsConnection) {
+
+	c := s.getOrAddEdsCluster(clusterName)
+	c.mutex.Lock()
+	existing := c.EdsClients[node]
+	c.mutex.Unlock()
+
+	// May replace an existing connection
+	if existing != nil {
+		existing.pushChannel <- false // force closing it
+	}
+	c.EdsClients[node] = connection
+}
+
+// getEdsCluster returns a cluster.
+func (s *DiscoveryServer) getEdsCluster(clusterName string) *EdsCluster {
+	// separate method only to have proper lock.
+	edsClusterMutex.Lock()
```

Comment:
RLock and RUnlock.

Reply:
Same as above... I don't mind requiring 1.10

Reply:
The mutex has RLock and RUnlock (I think the problem is that you are using Mutex and not RWMutex (https://golang.org/pkg/sync/#RWMutex) ). We actually have an example of this in pilot IIRC (jason added this long time back)

Reply:
We have plenty of RWMutexes in the codebase. No reason not to use one (I suggest above, actually). But, I think in this case especially (contested lock) we should use a [`sync.Map`](https://golang.org/pkg/sync/#Map) instead. It was basically designed for this use case (read heavy workload where concurrent writers operate on disjoint key sets).